agent/secrets/secrets.go

```diff
@@ -62,14 +63,20 @@ func (s *secrets) Reset() {
 type taskRestrictedSecretsProvider struct {
 	secrets   exec.SecretGetter
 	secretIDs map[string]struct{} // allow list of secret ids
+	taskID    string              // ID of the task the provider restricts for
 }
 
 func (sp *taskRestrictedSecretsProvider) Get(secretID string) (*api.Secret, error) {
 	if _, ok := sp.secretIDs[secretID]; !ok {
 		return nil, fmt.Errorf("task not authorized to access secret %s", secretID)
 	}
 
-	return sp.secrets.Get(secretID)
+	secret, err := sp.secrets.Get(identity.XorIDs(secretID, sp.taskID))
```

Comment:
Would this do?
```golang
	// First check if the secret is available with the task specific ID, the XOR of the original secret ID and the task ID.
	// That is the case when a secret driver has returned DoNotReuse == true for a secret value.
	secret, err := sp.secrets.Get(identity.XorIDs(secretID, sp.taskID))
	if err != nil {
		// Otherwise, which is the default case, the secret is retrieved by its original ID.
		return sp.secrets.Get(secretID)
	}
	// For all intents and purposes, the rest of the flow should deal with the original secret ID.
	secret.ID = secretID
	return secret, err
```